packages/flutter_tools/lib/src/localizations/gen_l10n.dart

```diff
@@ -123,23 +123,36 @@ String _syntheticL10nPackagePath(FileSystem fileSystem) => fileSystem.path.join(
 // For example, if placeholders are used for plurals and no type was specified, then the type will
 // automatically set to 'num'. Similarly, if such placeholders are used for selects, then the type
 // will be set to 'String'. For such placeholders that are used for both, we should throw an error.
-List<String> generateMethodParameters(Message message, bool useNamedParameters) {
-  return message.placeholders.values.map((Placeholder placeholder) {
+List<String> generateMethodParameters(Message message, LocaleInfo? locale, bool useNamedParameters) {
+
+  // Check the compatibility of template placeholders and locale placeholders.
+  final Map<String, Placeholder>? localePlaceholders = message.localePlaceholders[locale];
+
+  return message.templatePlaceholders.entries.map((MapEntry<String, Placeholder> e) {
+    final Placeholder placeholder = e.value;
+    final Placeholder? localePlaceholder = localePlaceholders?[e.key];
+    if (localePlaceholder != null && placeholder.type != localePlaceholder.type) {
+      throw L10nException(
```

Comment:
Note for second reviewer:

This change _could_[^1] break peoples' builds. Consider

```json
// app_en.arb
{
  "@@locale": "en",
  "springBegins": "Spring begins on {springStartDate}",
  "@springBegins": {
    "description": "The first day of spring",
    "placeholders": {
      "springStartDate": {
        "type": "DateTime",
        "format": "MMMMd"
      }
    }
}
```

and
```json
// app_ja.arb
{
  "@@locale": "ja",
  "springBegins": "春が始まるのは{springStartDate}",
  "@springBegins": {
    "description": "The first day of spring",
    "placeholders": {
      "springStartDate": {
        "type": "int", // Type change.
        "format": "compactCurrency"
      }
    }
  },
```

Then, trying to regenerate localization code will result in a failure:

```
Generating synthetic localizations package failed with 1 error:

Error: The placeholder, springStartDate, has its "type" resource attribute set to the "int" type in locale "ja", but it is "DateTime" in the template placeholders. For
compatibility with template placeholders, change the "type" attribute to "DateTime".
```

I think this failing the build makes sense here, and I believe there's no pre-existing behavior for users to be dependent on here (the `springStartDate` placeholder in the `ja` locale would have been silently ignored, hence this PR). I'm content with this landing without a breaking change announcement.

[^1]: And therefore _will_ 🙃

Reply:
I think that's reasonable.